h/static/scripts/annotation-sync.coffee

```diff
@@ -0,0 +1,190 @@
+      delete @cache[annotation.$$tag]
+      @_emit('annotationDeleted', [annotation])
+      res = this._format annotation
+      delete @cache[annotation.$$tag]
+      res
+
+    'sync': (ctx, annotations) ->
+      (this._format (this._parse a) for a in annotations)
+
+    'loadAnnotations': (txn, annotations) ->
+      annotations = (this._parse a for a in annotations)
+      @_emit('loadAnnotations', annotations)
+
+  # Handlers for events coming from this frame, to send them across the channel
+  _eventListeners:
+    'beforeAnnotationCreated': (event, annotation) ->
+      return if annotation.$$tag?
+      this._mkCallRemotelyAndParseResults('beforeCreateAnnotation')(annotation)
+      this
+
+    'annotationCreated': (event, annotation) ->
+      return unless annotation.$$tag? and @cache[annotation.$$tag]
+      this._mkCallRemotelyAndParseResults('createAnnotation')(annotation)
+      this
+
+    'annotationUpdated': (event, annotation) ->
+      return unless annotation.$$tag? and @cache[annotation.$$tag]
+      this._mkCallRemotelyAndParseResults('updateAnnotation')(annotation)
+      this
+
+    'annotationDeleted': (event, annotation) ->
+      return unless annotation.$$tag? and @cache[annotation.$$tag]
+      onFailure = (err) =>
+        delete @cache[annotation.$$tag] unless err
+      this._mkCallRemotelyAndParseResults('deleteAnnotation', onFailure)(annotation)
+      this
+
+    'annotationsLoaded': (event, annotations) ->
+      annotations = (this._format a for a in annotations when not a.$$tag)
+      return unless annotations.length
+      this._notify
+        method: 'loadAnnotations'
+        params: annotations
+      this
+
+  _syncCache: (channel) ->
+    # Synchronise (here to there) the items in our cache
+    annotations = (this._format a for t, a of @cache)
+    if annotations.length
+      channel.notify
+        method: 'loadAnnotations'
+        params: annotations
+
+  _mkCallRemotelyAndParseResults: (method, callBack) ->
+    (annotation) =>
+      # Wrap the callback function to first parse returned items
+      wrappedCallback = (failure, results) =>
+        unless failure?
+          this._parseResults results
+        callBack? failure, results
+
+      # Call the remote method
+      options =
+        method: method
+        callback: wrappedCallback
+        params: this._format(annotation)
+      @bridge.call(options)
+
+  # Parse returned annotations to update cache with any changes made remotely
+  _parseResults: (results) ->
+    for annotations in results
+      if Array.isArray(annotations)
+        for a in annotations
+          this._parse(a) unless a is null
+      else
+        # 'annotations' is in fact just a single annotation
+        this._parse(annotations) unless annotations is null
+    return
+
+  # Assign a non-enumerable tag to objects which cross the bridge.
+  # This tag is used to identify the objects between message.
+  _tag: (msg, tag) ->
+    return msg if msg.$$tag
```

Comment:
When exactly does this happen? And is it certain that `@cache` does not need updating in that situation?